docs/architecture.md

```diff
@@ -0,0 +1,473 @@
+}
+
+trait Relayer {
+    // wait until the relayer eth node is synced
+    async fn await_synced() -> Result<()>;
+    // wait until the relayer reaches a desired da_height
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+    // get the consensus key associated with a da_height
+    async fn get_poa_key(da_height: DaHeight) -> Result<ConsensusKey>;
+}
+
+trait Database {
+    // get the current fuel chain height
+    fn get_current_height() -> Result<BlockHeight>;
+    // insert consensus information associated with a sealed block that was locally produced
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+trait PeerToPeer {
+    // broadcast a newly produced block to the network
+    async fn broadcast_new_block(block: SealedFuelBlock) -> Result<()>;
+}
+```
+#### Adapters: fuel_core::service::adapters::poa_consensus
+```rust
+impl fuel_sync::ports::Consensus for Service<PoA> {
+    // insert impl here
+}
+```
+
+### Block Producer
+
+#### Ports: fuel_block_producer::ports
+
+```rust
+trait TransactionPool {
+    fn select_block_txs(max_gas: u64, block_height: BlockHeight) -> Result<Vec<Arc<CheckedTransaction>>>;
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Executor {
+    async fn produce_block(block: &mut FuelBlock) -> Result<()>;
+}
+
+trait Relayer {
+    fn get_finalized_da_height() -> Result<DaHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::block_producer
+```rust
+impl fuel_poa_consensus::ports::BlockProducer for Service<BlockProducer> {
+    // insert impl here
+}
+```
+
+### Block Importer
+
+#### Ports: fuel_block_importer::ports
+
+```rust
+trait Relayer {
+    // wait until the relayer is synced with ethereum
+    async fn await_synced() -> Result<()>;
+    // wait until a specific da_height is reached
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+}
+
+trait Executor {
+    async fn validate_and_store_block(block: &FuelBlock) -> Result<()>;
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+
+```rust
+trait PeerToPeer {
+    type SealedHeaderResponse: NetworkData<SealedFuelBlockHeader>;
+    type BlockResponse: NetworkData<Vec<SealedFuelBlock>>;
+    type GossipedBlockHeader: NetworkData<SealedFuelBlockHeader>;
+    
+    async fn fetch_best_network_block_header() -> Result<Self::SealedHeaderResponse>;
```

Comment:
Is this always guaranteed to return a response or should it be `Option<Result<Self::SealedHeaderResponse>>`?

Reply:
Also does this timeout / if so do we want to retry (so it would need to be an error case).

Reply:
This should be an Option if you're not connected to any peers

Reply:
Is it also possible that none of the peers have any sealed blocks?

Reply:
sure, that should also be a `None`